src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -371,151 +371,160 @@ internal unsafe void Start()
+                            break;
                         }
-                        if (((eventType & WatcherChangeTypes.Deleted) > 0))
+                        else if (handledRenameEvents == i)
                         {
-                            watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, relativePath);
+                            // If this event is the second in a rename pair then skip it
+                            continue;
                         }
-                        if (((eventType & WatcherChangeTypes.Renamed) > 0))
+                        else if (CheckIfPathIsNested(path) && ((eventType = FilterEvents(eventFlags[i])) != 0))
                         {
-                            // Find the rename that is paired to this rename, which should be the next rename in the list
-                            int pairedId = FindRenameChangePairedChange(i, eventFlags, numEvents);
-                            if (pairedId == int.MinValue)
+                            // The base FileSystemWatcher does a match check against the relative path before combining with
+                            // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
+                            ReadOnlySpan<char> relativePath = ReadOnlySpan<char>.Empty;
+                            if (path.Length > _fullDirectory.Length && path.StartsWith(_fullDirectory, StringComparison.OrdinalIgnoreCase))
                             {
-                                // Getting here means we have a rename without a pair, meaning it should be a create for the
-                                // move from unwatched folder to watcher folder scenario or a move from the watcher folder out.
-                                // Check if the item exists on disk to check which it is
-                                // Don't send a new notification if we already sent one for this event.
-                                if (DoesItemExist(path, IsFlagSet(eventFlags[i], FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile)))
-                                {
-                                    if ((eventType & WatcherChangeTypes.Created) == 0)
-                                    {
-                                        watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Created, relativePath);
-                                    }
-                                }
-                                else if ((eventType & WatcherChangeTypes.Deleted) == 0)
-                                {
-                                    watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, relativePath);
-                                }
+                                // Remove the root directory to get the relative path
+                                relativePath = path[_fullDirectory.Length..];
```

Comment:
Nit: The previous Slice call is ever so slightly better from a performance perspective, though here it's unlikely to matter.  The difference is that the range index syntax is going to map to using Slice(start, count), whereas the previous call was just using Slice(start), and the latter has less argument validation to do.

Reply:
yeah, and there is actually no readability benefit compared to the slice.